subsys/bootloader/provision/provision_flash.c

```diff
@@ -99,3 +104,90 @@ void invalidate_public_key(u32_t key_idx)
 		nrfx_nvmc_word_write((u32_t)invalidation_token, INVALID_VAL);
 	}
 }
+
+
+static u16_t read_halfword(u16_t *ptr)
+{
+	bool top_half = ((u32_t)ptr % 4); /* Addr not div by 4 */
+	u32_t target_addr = (u32_t)ptr & ~3; /* Floor address */
+
+	u32_t val32 = *(u32_t *)target_addr;
+	u16_t val16 = val32 & 0x0000FFFF;
+	u16_t val16_shifted = (val32 >> 16) & 0x0000FFFF;
+
+	return top_half ? val16_shifted : val16;
+}
+
+
+static void write_halfword(u16_t *ptr, u16_t val)
+{
+	bool top_half = ((u32_t)ptr % 4); /* Addr not div by 4 */
+	u32_t target_addr = (u32_t)ptr & ~3; /* Floor address */
+
+	u32_t val32 = (u32_t)val | 0xFFFF0000;
+	u32_t val32_shifted = ((u32_t)val << 16) | 0x0000FFFF;
+
+	nrfx_nvmc_word_write(target_addr, top_half ? val32_shifted : val32);
+}
+
+
+static struct monotonic_counter *get_counter_struct(void)
+{
+	return (struct monotonic_counter *)
+		&p_provision_data->key_data[num_public_keys_read()];
+}
+
+
+u32_t num_monotonic_counter_slots(void)
+{
+	u32_t num_slots = get_counter_struct()->num_counter_slots;
+
+	return num_slots != 0xFFFFFFFF ? num_slots : 0;
+}
+
+
+static u16_t *get_monotonic_counter_ptr(u16_t *outval)
+{
+	*outval = 0;
+	u16_t *slots = get_counter_struct()->counter_slots;
+
+	for (u32_t i = 0; (i < num_monotonic_counter_slots()); i++) {
+		u16_t counter = read_halfword(&slots[i]);
+		if (counter == 0xFFFF) {
+			return &slots[i];
+		}
+		if (*outval < counter) {
```

Comment:
Should we check and fail here if we find a non-increasing value?

Reply:
This merits more discussion. I thought about it as I was making it, and decided to be lenient, since no-one wants a bricked device by accident, and I don't see any security implications. I'll think/discuss some more. I added it to the list of open questions at the top.